File: src/server/orchestrator/safeguardWatcher.ts
import fs from 'node:fs';
import path from 'node:path';
import { SafeguardRepo } from './safeguardRepo.js';
import type { FilesPayload } from '../../shared/types.js';

export interface SafeguardWatcherOptions {
  bareDir: string;
  branch: string;
  onSnapshot: (s: FilesPayload) => void;
  onError?: (reason: string) => void;
}

export class SafeguardWatcher {
  private watcher: fs.FSWatcher | null = null;
  private packedWatcher: fs.FSWatcher | null = null;
  private readonly repo: SafeguardRepo;
  private lastSha: string | null = null;
  private started = false;

  constructor(private readonly opts: SafeguardWatcherOptions) {
    this.repo = new SafeguardRepo(opts.bareDir);
  }

  async start(): Promise<void> {
    if (this.started) return;
    this.started = true;
    await this.emit();
    const refDir = path.join(this.opts.bareDir, 'refs', 'heads');
    fs.mkdirSync(refDir, { recursive: true });
    try {
      this.watcher = fs.watch(refDir, { recursive: false }, () => {
        void this.emit();
      });
    } catch (e) {
      this.opts.onError?.(String(e));
    }
    const packedPath = path.join(this.opts.bareDir, 'packed-refs');
    try {
      this.packedWatcher = fs.watch(path.dirname(packedPath), (_ev, fn) => {
        if (fn === 'packed-refs') void this.emit();
      });
    } catch (e) {
      this.opts.onError?.(String(e));
    }
  }

  async stop(): Promise<void> {
    this.started = false;
    this.watcher?.close(); this.watcher = null;
    this.packedWatcher?.close(); this.packedWatcher = null;
  }

  private async emit(): Promise<void> {
    const head = this.repo.head(this.opts.branch);
    const sha = head?.sha ?? null;
    if (sha === this.lastSha) return;
    this.lastSha = sha;
    const headFiles = this.repo.headFiles(this.opts.branch);
    const payload: FilesPayload = {
      dirty: [],
      head,
      headFiles,
      branchBase: null,
      live: false,
      dirty_submodules: [],
    };
    this.opts.onSnapshot(payload);
  }
}

File: server-elixir/lib/fbi/orchestrator/safeguard_watcher.ex
defmodule FBI.Orchestrator.SafeguardWatcher do
  @moduledoc "Port of src/server/orchestrator/safeguardWatcher.ts."
  use GenServer

  alias FBI.Orchestrator.SafeguardRepo

  defstruct [:bare_dir, :branch, :on_snapshot, :last_sha, :fs_pid]

  def start_link(opts) do
    GenServer.start_link(__MODULE__, opts)
  end

  def stop(pid), do: GenServer.stop(pid)

  @impl true
  def init(opts) do
    state = %__MODULE__{
      bare_dir: Keyword.fetch!(opts, :bare_dir),
      branch: Keyword.fetch!(opts, :branch),
      on_snapshot: Keyword.fetch!(opts, :on_snapshot),
      last_sha: nil,
      fs_pid: nil
    }

    {:ok, fs_pid} = FileSystem.start_link(dirs: [state.bare_dir])

    FileSystem.subscribe(fs_pid)

    state = %{state | fs_pid: fs_pid}
    state = emit(state)
    {:ok, state}
  end

  @impl true
  def handle_info({:file_event, _pid, {_path, _events}}, state) do
    state = emit(state)
    {:noreply, state}
  end

  @impl true
  def handle_info({:file_event, _pid, :stop}, state) do
    {:noreply, state}
  end

  @impl true
  def terminate(_reason, state) do
    if state.fs_pid, do: GenServer.stop(state.fs_pid, :normal)
    :ok
  end

  defp emit(state) do
    head = SafeguardRepo.head(state.bare_dir, state.branch)
    sha = if head, do: head.sha, else: nil

    if sha != state.last_sha do
      head_files = SafeguardRepo.head_files(state.bare_dir, state.branch)

      payload = %{
        dirty: [],
        head: head,
        head_files: head_files,
        branch_base: nil,
        live: false,
        dirty_submodules: []
      }

      state.on_snapshot.(payload)
      %{state | last_sha: sha}
    else
      state
    end
  end
end
